buildsystem/generator/golang.go

```diff
@@ -0,0 +1,597 @@
+	Name string `yaml:"name,omitempty"`
+	// Path is the path of the package relative to the root package. For
+	// example, "." refers to the root package while "util" refers to a
+	// subpackage called "util" within the root package.
+	Path string `yaml:"path"`
+	// Tags are labels that allow you to logically group related packages.
+	Tags    []string             `yaml:"tags,omitempty"`
+	Options GolangRuntimeOptions `yaml:"options,omitempty"`
+}
+
+func (gp *GolangPackage) Validate() error {
+	catcher := grip.NewBasicCatcher()
+	catcher.NewWhen(gp.Path == "", "need to specify package path")
+	tags := map[string]struct{}{}
+	for _, tag := range gp.Tags {
+		if _, ok := tags[tag]; ok {
+			catcher.Errorf("duplicate tag '%s'", tag)
+		}
+		tags[tag] = struct{}{}
+	}
+	catcher.Add(gp.Options.Validate())
+	return catcher.Resolve()
+}
+
+// GolangVariant defines a mapping between distros that run packages and the
+// golang packages to run.
+type GolangVariant struct {
+	Name    string   `yaml:"name"`
+	Distros []string `yaml:"distros"`
+	// Packages lists the names or paths of the package relative to the root
+	// package, which can be one of the following:
+	// - A name of a package explicitly listed in the packages section.
+	// - A path of a package explicitly listed in the packages section.
+	// - A path of a package that matches a path to a directory containing tests
+	//   relative to the root package.
+	// - A package tag, beginning with a period (e.g. ".tag").
+	Packages []VariantPackage `yaml:"packages"`
+	// Options are variant-specific options that modify test execution.
+	// Explicitly setting these values will override options specified under the
+	// definitions of packages.
+	Options *GolangRuntimeOptions `yaml:"options,omitempty"`
+}
+
+func (gv *GolangVariant) Validate() error {
+	catcher := grip.NewBasicCatcher()
+	catcher.NewWhen(gv.Name == "", "missing variant name")
+	catcher.NewWhen(len(gv.Distros) == 0, "need to specify at least one distro")
+	catcher.NewWhen(len(gv.Packages) == 0, "need to specify at least one package")
+	pkgPaths := map[string]struct{}{}
+	pkgNames := map[string]struct{}{}
+	pkgTags := map[string]struct{}{}
+	for _, pkg := range gv.Packages {
+		catcher.Wrap(pkg.Validate(), "invalid package reference")
+		if path := pkg.Path; path != "" {
+			if _, ok := pkgPaths[path]; ok {
+				catcher.Errorf("duplicate reference to package path '%s'", path)
+			}
+			pkgPaths[path] = struct{}{}
+		}
+		if name := pkg.Name; name != "" {
+			if _, ok := pkgNames[name]; ok {
+				catcher.Errorf("duplicate reference to package name '%s'", name)
+			}
+			pkgNames[name] = struct{}{}
+		}
+		if tag := pkg.Tag; tag != "" {
+			if _, ok := pkgTags[tag]; ok {
+				catcher.Errorf("duplicate reference to package tag '%s'", tag)
+			}
+			pkgTags[tag] = struct{}{}
+		}
+	}
+	if gv.Options != nil {
+		catcher.Wrap(gv.Options.Validate(), "invalid runtime options")
+	}
+	return catcher.Resolve()
+}
+
+type VariantPackage struct {
```

Comment:
I'd put all the tags on this just be be safe.

Reply:
The tags must go on the package itself. This is just used in the variant definition to refer to a package definition by name/path/tag.